lib/offsite_payments/integrations/paytm.rb

```diff
@@ -0,0 +1,317 @@
+            @fields[field].gsub!(/[^a-zA-Z0-9\-_@\/\s.]/, '') if @fields[field]
+          end
+        end
+
+      end
+
+      class Notification < OffsitePayments::Notification
+	  
+		PaytmResponseParams = ['SUBS_ID','MID','BANKTXNID','TXNAMOUNT','CURRENCY','STATUS','RESPCODE','RESPMSG','TXNDATE','GATEWAYNAME','BANKNAME','PAYMENTMODE','PROMO_CAMP_ID','PROMO_STATUS','PROMO_RESPCODE','ORDERID','TXNID','REFUNDAMOUNT','REFID']
+        def initialize(post, options = {})
+          super(post, options)
+          @merchant_id = options[:credential1]
+          @secret_key = options[:credential2]
+        end
+
+        def complete?
+          status == "Completed"
+        end
+
+        def status
+          case transaction_status
+          when 'TXN_SUCCESS' then 'Completed'
+          when 'TXN_FAILURE' then 'Failed'
+          when 'pending' then 'Pending'
+          end
+        end
+
+        def invoice_ok?( order_id )
+          order_id.to_s == invoice.to_s
+        end
+
+        # Order amount should be equal to gross - discount
+        def amount_ok?( order_amount, order_discount = BigDecimal.new( '0.0' ) )
+          BigDecimal.new( original_gross ) == order_amount
+        end
+
+        # Status of transaction return from the Paytm. List of possible values:
+        # <tt>SUCCESS</tt>::
+        # <tt>PENDING</tt>::
+        # <tt>FAILURE</tt>::
+        def transaction_status
+          params['STATUS']
+        end
+
+        # ID of this transaction (Paytm transaction id)
+        def transaction_id
+          params['TXNID']
+        end
+
+        # Mode of Payment
+        #
+        # 'CC' for credit-card
+        # 'NB' for net-banking
+        
+        def type
+          params['PAYMENTMODE']
+        end
+
+        # What currency have we been dealing with
+        def currency
+          'INR'
+        end
+
+        def item_id
+          params['ORDERID']
+        end
+
+        # This is the invoice which you passed to Paytm
+        def invoice
+          params['ORDERID']
+        end
+
+        # Merchant Id provided by the Paytm
+        def account
+          params['MID']
+        end
+
+        # original amount send by merchant
+        def original_gross
+          params['TXNAMOUNT']
+        end
+
+        def gross
+          parse_and_round_gross_amount(params['TXNAMOUNT'])
+        end
+
+               
+
+        def checksum
+          params['CHECKSUMHASH'].gsub(" ","+")
+        end
+
+        def message
+          @message || params['error']
+        end
+
+        def acknowledge(authcode = nil)
+          checksum_ok?
+        end
+
+        def checksum_ok?
+			
+			check_sum = checksum.gsub("\n","")
+			check_sum = check_sum.gsub(" ","+")
+			
+			if check_sum.nil?
+			  return false
+			end
+			
+			generated_check_sum = nil			
+			aes = OpenSSL::Cipher::Cipher.new("aes-128-cbc")
+			aes.decrypt
+			aes.key = @secret_key
+			aes.iv = '@@@@&&&&####$$$$'
```

Comment:
Hi Alexander,

I have check regarding your concern and below are my findings.

Dictionary attack is only possible issue with fixed iv AES (CBC mode), but
as per paytm logic we add a random 4 digit salt at the end of request
string before AES (CBC mode) encryption and this salt is not present in
request, So with paytm checksum algorithim possibility of dictionary
attacks is nil and we can use fixed iv without any security concern.

Note: As per Paytm Checksum algorithm we need fixed iv we can't modify it.

Regards,

On Wed, Jun 15, 2016 at 10:11 PM, Alexander Profeit <
notifications@github.com> wrote:

> In lib/offsite_payments/integrations/paytm.rb
> <https://github.com/activemerchant/offsite_payments/pull/200#discussion_r67198886>
> :
>
> > +        end
> > +
> > +        def checksum_ok?
> > +			
> > +			check_sum = checksum.gsub("\n","")
> > +			check_sum = check_sum.gsub(" ","+")
> > +			
> > +			if check_sum.nil?
> > +			  return false
> > +			end
> > +			
> > +			generated_check_sum = nil			
> > +			aes = OpenSSL::Cipher::Cipher.new("aes-128-cbc")
> > +			aes.decrypt
> > +			aes.key = @secret_key
> > +			aes.iv = '@@@@&&&&####$$$$'
>
> I still see this fixed iv here, is this going to be updated as well? This
> should be randomly generated for each encrypt, then normally padded on the
> beginning of the encrypted data.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/activemerchant/offsite_payments/pull/200/files/705da561badcb2dee3fb177edb7474f9905ff37d#r67198886>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AL1PqBBfr2NHp_DxIS9w7A0cgcWJxe3aks5qMCtEgaJpZM4IQnF4>
> .
>



-- 
*Mohit Aggarwal*
Senior Software Engineer
+91 8826036707
[image: Inline image 1]
www.Paytm.com

-- 


Disclaimer :-

This email and any files transmitted with it are confidential and 
intended solely for the use of the individual or entity to which they are 
addressed. If you are not the intended recipient, you should not 
disseminate, distribute or copy this e-mail. Please notify the sender 
immediately by e-mail and destroy all copies of this message and any 
attachments. Any views or opinions presented in this email are solely those 
of the author and do not necessarily represent those of the company. 

Warning: Although the company has taken reasonable precautions to ensure no 
viruses are present in this email, the company cannot accept responsibility 
for any loss or damage arising from the use of this email or attachments.